src/util/thread_pool.hpp

```diff
@@ -0,0 +1,41 @@
+#pragma once
+
+#include <boost/asio.hpp>
+
+#include "functional.hpp"
+
+namespace taraxa::util {
+
+class ThreadPool : std::enable_shared_from_this<ThreadPool> {
+  boost::asio::io_context ioc_;
+  boost::asio::executor_work_guard<decltype(ioc_)::executor_type> ioc_work_;
+  std::vector<std::thread> threads_;
+  std::mutex threads_mu_;
+
+ public:
+  explicit ThreadPool(size_t num_threads, bool _start = true);
+  ~ThreadPool();
+
+  ThreadPool(ThreadPool const &) = delete;
+  ThreadPool &operator=(ThreadPool const &) = delete;
+
+  auto &unsafe_get_io_context() { return ioc_; }
+
+  void start();
+  void stop();
+
+  template <typename... T>
+  auto post(T &&... args) {
+    return boost::asio::post(ioc_, std::forward<T>(args)...);
+  }
```

Comment:
Why put half of the class methods definitions into the cpp and half leave in header ?

Reply:
The templated methods should be in headers (read the basics to find why).
For the rest, I kind of agree, except I think leaving one-liners in headers is not too bad (depends on how much template expansions that might trigger).
The answer to your question - because it's not a huge problem, and because it's more convenient to work with logically same code in one file than in two files, I usually first write everything in header, but then sometimes forget to extract to .cpp file when I'm done with it.